File: src/exercise-1.16/answer-1.16.rb
#!/usr/bin/ruby -w

require 'test/unit'

=begin
Ruby solution to SICP exercise 1.16
=end

def exponential(base, exp)
  result = 1

  while exp > 0
    if exp.even?
      base *= base
      exp  /= 2
    else
      result *= base
      exp -= 1
    end
  end

  return result
end


class TestExponential < Test::Unit::TestCase
  def test_exponential
    (0..10).each do |i|
      (0..10).each do |j|
        assert_equal(i**j, exponential(i, j))
      end
    end
  end
end

File: src/exercise-1.16/answer-1.16.clj
(comment "Clojure solution to SICP exercise 1.16")

(ns sicp-1.16
  (:use [clojure.test]))

(defn exponential [base, exp]
  (defn square [x] (* x x))
  (defn exp-iter [b n result]
    (cond
      (= n 0)   result
      (even? n) (exp-iter (* b b) (/ n 2) result)
      :else     (exp-iter b (- n 1) (* result b))))
  (exp-iter base exp 1))

(deftest test-exponential
  (is (= (exponential 1 1) 1))
  (is (= (exponential 2 2) 4))
  (is (= (exponential 3 3) 27))
  (is (= (exponential 4 4) 256))
  )

(run-tests)

